PR: https://github.com/odoo/odoo/pull/225252

From: b6bbc31f7464b0fdae6a83122b42b99cd7fca1cf
From: william-andre
Date: 2025-09-07 06:44:51

Breaking data model changes scores: del:2 + add:0, change matches:
-    tax_negate = fields.Boolean(string="Negate Tax Balance", help="Check this box to negate the absolute value of the balance of the lines associated with this tag in tax report computation.")
-    tax_tag_invert = fields.Boolean(

Total Changes: 150

[IMP] account: remove tax tag invert

`tax_tag_invert` and `tax_negate` are part of a useless blackbox that
prevents both users and devs of understanding the sign of the tax
report.
Removing those tags and forcing to create a new tag for cases where a
tax grid increases with both debit and credit makes it easier to audit
the tax report.

Part-of: odoo/odoo#225252
Related: odoo/enterprise#93761
Related: odoo/documentation#14474
Related: odoo/upgrade#8361
Signed-off-by: William André (wan) <wan@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_account_tag.py
+++ b/addons/account/models/account_account_tag.py
@@ -1,6 +1,6 @@
 from odoo import api, fields, models, _
 from odoo.fields import Domain
-from odoo.tools.sql import SQL
+from odoo.tools import SQL, Query
 from odoo.exceptions import UserError
 
 
@@ -12,8 +12,12 @@ class AccountAccountTag(models.Model):
     applicability = fields.Selection([('accounts', 'Accounts'), ('taxes', 'Taxes'), ('products', 'Products')], required=True, default='accounts')
     color = fields.Integer('Color Index')
     active = fields.Boolean(default=True, help="Set active to false to hide the Account Tag without removing it.")
-    tax_negate = fields.Boolean(string="Negate Tax Balance", help="Check this box to negate the absolute value of the balance of the lines associated with this tag in tax report computation.")
     country_id = fields.Many2one(string="Country", comodel_name='res.country', help="Country for which this tag is available, when applied on taxes.")
+    # If the tag was generated by a report line for the `tax_tags` engine, these fields allow getting
+    # the sign of the report line used to display the amount.
+    # If balance_negate is true then it means we use `-balance` instead of `balance`
+    report_expression_id = fields.Many2one('account.report.expression', compute='_compute_report_expression_id')
+    balance_negate = fields.Boolean(compute='_compute_report_expression_id')
 
     _name_uniq = models.Constraint(
         'unique(name, applicability, country_id)',
@@ -32,6 +36,37 @@ class AccountAccountTag(models.Model):
                 name = _("%(tag)s (%(country_code)s)", tag=tag.name, country_code=tag.country_id.code)
             tag.display_name = name
 
+    @api.depends('name')
+    def _compute_report_expression_id(self):
+        query = self._search([('id', 'in', self.ids)])
+        id2expression = {tag_id: vals for tag_id, *vals in self.env.execute_query(query.select(
+            SQL.identifier(query.table, 'id'),
+            self._field_to_sql(query.table, 'report_expression_id', query),
+            self._field_to_sql(query.table, 'balance_negate', query),
+        ))}
+        for tag in self:
+            tag.report_expression_id, tag.balance_negate = id2expression.get(tag._origin.id, (False, False))
+
+    def _field_to_sql(self, alias: str, field_expr: str, query: (Query | None) = None) -> SQL:
+        if field_expr in ('report_expression_id', 'balance_negate'):
+            rhs_alias = query.make_alias(alias, 'expression')
+            if rhs_alias not in query._tables:
+                query.add_join(
+                    kind='LEFT JOIN',
+                    alias=rhs_alias,
+                    table='account_report_expression',
+                    condition=SQL(
+                        "%s->>'en_US' = LTRIM(%s, '-')",
+                        SQL.identifier(alias, 'name'),
+                        SQL.identifier(rhs_alias, 'formula'),
+                    ),
+                )
+            if field_expr == 'report_expression_id':
+                return SQL.identifier(rhs_alias, 'id')
+            if field_expr == 'balance_negate':
+                return SQL("STARTS_WITH(%s, '-')", SQL.identifier(rhs_alias, 'formula'))
+        return super()._field_to_sql(alias, field_expr, query)
+
     @api.model_create_multi
     def create(self, vals_list):
         tags = super().create(vals_list)
@@ -50,32 +85,27 @@ class AccountAccountTag(models.Model):
         return rslt_tags.with_context(lang=original_lang)  # Restore original language, in case the name of the tags needs to be shown/modified
 
     @api.model
-    def _get_tax_tags_domain(self, tag_name, country_id, sign=None):
-        """ Returns a domain to search for all the tax tags corresponding to the tag name given in parameter
+    def _get_tax_tags_domain(self, formula, country_id):
+        """ Returns a domain to search for all the tax tags corresponding to the formula given in parameter
         in the specified country.
         """
-        escaped_tag_name = tag_name.replace('\\', '\\\\').replace('%', r'\%').replace('_', r'\_')
         return [
-            ('name', '=like', (sign or '_') + escaped_tag_name),
+            ('name', '=', formula.lstrip('-')),
             ('country_id', '=', country_id),
-            ('applicability', '=', 'taxes')
+            ('applicability', '=', 'taxes'),
         ]
 
     def _get_related_tax_report_expressions(self):
         if not self:
             return self.env['account.report.expression']
 
-        or_domains = []
-        for record in self:
-            expr_domain = [
-                '&',
-                ('report_line_id.report_id.country_id', '=', record.country_id.id),
-                ('formula', '=', record.name[1:]),
-            ]
-            or_domains.append(expr_domain)
-
-        domain = Domain('engine', '=', 'tax_tags') & Domain.OR(or_domains)
-        return self.env['account.report.expression'].search(domain)
+        return self.env['account.report.expression'].search(Domain('engine', '=', 'tax_tags') & Domain.OR(
+            (
+                Domain('report_line_id.report_id.country_id', '=', record.country_id.id)
+                & Domain('formula', 'in', (record.name, '-' + record.name))
+            )
+            for record in self
+        ))
 
     @api.ondelete(at_uninstall=False)
     def _unlink_except_master_tags(self):

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -5002,7 +5002,6 @@ class AccountMove(models.Model):
                         **vals,
                         'amount_currency': 0.0,
                         'balance': 0.0,
-                        'display_type': 'epd',  # Used to compute tax_tag_invert for early payment discount lines
                     })
                     line_vals['amount_currency'] += vals['amount_currency']
                     line_vals['balance'] += vals['balance']

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -237,13 +237,6 @@ class AccountMoveLine(models.Model):
     # Technical field holding custom data for the taxes computation engine.
     extra_tax_data = fields.Json()
 
-    # Technical field. True if the balance of this move line needs to be
-    # inverted when computing its total for each tag (for sales invoices, for # example)
-    tax_tag_invert = fields.Boolean(
-        string="Invert Tags",
-        compute='_compute_tax_tag_invert', store=True, readonly=False, copy=False,
-    )
-
     # === Reconciliation fields === #
     amount_residual = fields.Monetary(
         string='Residual Amount',
@@ -851,28 +844,6 @@ class AccountMoveLine(models.Model):
         for line in self:
             line.allowed_uom_ids = line.product_id.uom_id | line.product_id.uom_ids
 
-    @api.depends('move_id.move_type', 'tax_ids', 'tax_repartition_line_id', 'debit', 'credit', 'tax_tag_ids', 'is_refund',
-                 'move_id.tax_cash_basis_origin_move_id')
-    def _compute_tax_tag_invert(self):
-        for record in self:
-            origin_move_id = record.move_id.tax_cash_basis_origin_move_id or record.move_id
-            if not record.tax_repartition_line_id and not record.tax_ids:
-                # Invoices imported from other softwares might only have kept the tags, not the taxes.
-                record.tax_tag_invert = record.tax_tag_ids and origin_move_id.is_inbound()
-
-            elif origin_move_id.move_type == 'entry':
-                # For misc operations, cash basis entries and write-offs from the bank reconciliation widget
-                tax = record.tax_repartition_line_id.tax_id or record.tax_ids[:1]
-                is_refund = record.is_refund
-                tax_type = tax.type_tax_use
-                if record.display_type == 'epd':  # In case of early payment, tax_tag_invert is independent of the balance of the line
-                    record.tax_tag_invert = tax_type == 'purchase'
-                else:
-                    record.tax_tag_invert = (tax_type == 'purchase' and is_refund) or (tax_type == 'sale' and not is_refund)
-            else:
-                # For invoices with taxes
-                record.tax_tag_invert = origin_move_id.is_inbound()
-
     @api.depends('product_id')
     def _compute_product_uom_id(self):
         for line in self.filtered(lambda l: l.parent_state == 'draft'):

--- a/addons/account/models/account_partial_reconcile.py
+++ b/addons/account/models/account_partial_reconcile.py
@@ -413,7 +413,6 @@ class AccountPartialReconcile(models.Model):
             'partner_id': tax_line.partner_id.id,
             'analytic_distribution': tax_line.analytic_distribution,
             'display_type': tax_line.display_type,
-            # No need to set tax_tag_invert as on the base line; it will be computed from the repartition line
         }
 
     @api.model

--- a/addons/account/models/account_report.py
+++ b/addons/account/models/account_report.py
@@ -657,10 +657,9 @@ class AccountReportExpression(models.Model):
             vals['formula'] = re.sub(r'\s+', ' ', vals['formula'].strip())
 
     def _create_tax_tags(self, tag_name, country):
-        existing_tags = self.env['account.account.tag']._get_tax_tags(tag_name, country.id)
-        if len(existing_tags) < 2:
-            # We can have only one tag in case we archived it and deleted its unused complement sign
-            tag_vals = self._get_tags_create_vals(tag_name, country.id, existing_tag=existing_tags)
+        existing_tag = self.env['account.account.tag']._get_tax_tags(tag_name, country.id)
+        if not existing_tag:
+            tag_vals = self._get_tags_create_vals(tag_name, country.id, existing_tag=existing_tag)
             self.env['account.account.tag'].create(tag_vals)
 
     @api.model_create_multi
@@ -709,7 +708,6 @@ class AccountReportExpression(models.Model):
             former_formulas_by_country[expr.report_line_id.report_id.country_id].append(expr.formula)
 
         result = super().write(vals)
-
         for country, former_formulas_list in former_formulas_by_country.items():
             for former_formula in former_formulas_list:
                 new_tax_tags = self.env['account.account.tag']._get_tax_tags(vals['formula'], country.id)
@@ -720,13 +718,7 @@ class AccountReportExpression(models.Model):
 
                     if former_tax_tags and all(tag_expr in self for tag_expr in former_tax_tags._get_related_tax_report_expressions()):
                         # If we're changing the formula of all the expressions using that tag, rename the tag
-                        positive_tags, negative_tags = former_tax_tags.sorted(lambda x: x.tax_negate)
-                        if self.pool['account.tax'].name.translate:
-                            positive_tags._update_field_translations('name', {'en_US': f"+{vals['formula']}"})
-                            negative_tags._update_field_translations('name', {'en_US': f"-{vals['formula']}"})
-                        else:
-                            positive_tags.name = f"+{vals['formula']}"
-                            negative_tags.name = f"-{vals['formula']}"
+                        former_tax_tags._update_field_translations('name', {'en_US': vals['formula']})
                     else:
                         # Else, create a new tag. Its the compute functions will make sure it is properly linked to the expressions
                         tag_vals = self.env['account.report.expression']._get_tags_create_vals(vals['formula'], country.id)
@@ -746,7 +738,7 @@ class AccountReportExpression(models.Model):
         for tag in expressions_tags:
             other_expression_using_tag = self.env['account.report.expression'].sudo().search([
                 ('engine', '=', 'tax_tags'),
-                ('formula', '=', tag.with_context(lang='en_US').name[1:]),  # we escape the +/- sign
+                ('formula', '=', tag.with_context(lang='en_US').name),
                 ('report_line_id.report_id.country_id', '=', tag.country_id.id),
                 ('id', 'not in', self.ids),
             ], limit=1)
@@ -854,7 +846,7 @@ class AccountReportExpression(models.Model):
 
         return totals_by_code
 
-    def _get_matching_tags(self, sign=None):
+    def _get_matching_tags(self):
         """ Returns all the signed account.account.tags records whose name matches any of the formulas of the tax_tags expressions contained in self.
         """
         tag_expressions = self.filtered(lambda x: x.engine == 'tax_tags')
@@ -864,34 +856,20 @@ class AccountReportExpression(models.Model):
         or_domains = []
         for tag_expression in tag_expressions:
             country = tag_expression.report_line_id.report_id.country_id
-            or_domains.append(self.env['account.account.tag']._get_tax_tags_domain(tag_expression.formula, country.id, sign))
+            or_domains.append(self.env['account.account.tag']._get_tax_tags_domain(tag_expression.formula, country.id))
 
         return self.env['account.account.tag'].with_context(active_test=False, lang='en_US').search(Domain.OR(or_domains))
 
     @api.model
     def _get_tags_create_vals(self, tag_name, country_id, existing_tag=None):
-        """
-        We create the plus and minus tags with tag_name.
-        In case there is an existing_tag (which can happen if we deleted its unused complement sign)
-        we only recreate the missing sign.
-        """
-        minus_tag_vals = {
-          'name': '-' + tag_name,
-          'applicability': 'taxes',
-          'tax_negate': True,
-          'country_id': country_id,
-        }
-        plus_tag_vals = {
-          'name': '+' + tag_name,
+        tag_vals = {
+          'name': tag_name.lstrip('-'),
           'applicability': 'taxes',
-          'tax_negate': False,
           'country_id': country_id,
         }
         res = []
-        if not existing_tag or not existing_tag.tax_negate:
-            res.append(minus_tag_vals)
-        if not existing_tag or existing_tag.tax_negate:
-            res.append(plus_tag_vals)
+        if not existing_tag:
+            res.append(tag_vals)
         return res
 
     def _get_carryover_target_expression(self, options):

--- a/addons/account/models/account_tax.py
+++ b/addons/account/models/account_tax.py
@@ -1556,9 +1556,6 @@ class AccountTax(models.Model):
             # If the document is a refund or not to know which repartition lines must be used.
             'is_refund': load('is_refund', False),
 
-            # If the tags must be inverted or not.
-            'tax_tag_invert': load('tax_tag_invert', False),
-
             # Extra fields for tax lines generation:
             'partner_id': load('partner_id', self.env['res.partner']),
             'account_id': load('account_id', self.env['account.account']),
@@ -2927,7 +2924,6 @@ class AccountTax(models.Model):
         base_lines_to_update = []
         for base_line in base_lines:
             sign = base_line['sign']
-            tax_tag_invert = base_line['tax_tag_invert']
             tax_details = base_line['tax_details']
             base_lines_to_update.append((
                 base_line,
@@ -2943,7 +2939,7 @@ class AccountTax(models.Model):
                     grouping_key = frozendict(tax_rep_data['grouping_key'])
                     tax_line = tax_lines_mapping[grouping_key]
                     tax_line['name'] = base_line.get('manual_tax_line_name', tax.name)
-                    tax_line['tax_base_amount'] += sign * tax_data['base_amount'] * (-1 if tax_tag_invert else 1)
+                    tax_line['tax_base_amount'] += sign * tax_data['base_amount']
                     tax_line['amount_currency'] += sign * tax_rep_data['tax_amount_currency']
                     tax_line['balance'] += sign * tax_rep_data['tax_amount']
 

--- a/addons/account/models/chart_template.py
+++ b/addons/account/models/chart_template.py
@@ -1216,7 +1216,8 @@ class AccountChartTemplate(models.AbstractModel):
         def mapping_getter(*args):
             res = []
             for tag in args:
-                if re.match(r"^\w+\.\w+$", tag):
+                # make sure that it is a xmlid and not a random tag containing a `.` by checking the module name exists
+                if (match := re.match(r"^(?P<module>\w+)\.\w+$", tag)) and self.env['ir.module.module']._get(match.group('module')):
                     # xml_id => explicit data, doesn't need to be mapped
                     res.append(tag)
                 else:
